src/misc/umap.c

```diff
@@ -543,7 +543,12 @@ static igraph_error_t igraph_apply_forces(const igraph_t *graph,  const igraph_v
             continue;
         }
 
-        IGRAPH_CHECK(igraph_edge(graph, eid, &from, &to));
+        /* half the time, swap the from/to, otherwise some vertices are never moved */
+        if (RNG_UNIF01() > 0.5) {
+            IGRAPH_CHECK(igraph_edge(graph, eid, &from, &to));
```

Comment:
Just to be aware, `igraph_edge()` can potentially be noticeably slower than `IGRAPH_FROM` and `IGRAPH_TO`. I would prefer to avoid it in new code, and personally I'd rather remove this function eventually.

Reply:
Ah yes, now I remember that `igraph_edge()` is exposed in high-level interfaces, so it has to stay. But we have faster alternatives for internal use.

Reply:
yep, now using the macros, thanks